gitlabform/processors/project/project_job_token_scope_processor.py

```diff
@@ -0,0 +1,186 @@
+
+    def _process_configuration(self, project_and_group: str, configuration: dict):
+        job_token_config = configuration.get("job_token_scope", {})
+        verbose(f"Job Token Scope config: {job_token_config}")
+
+        project = self.gl.projects.get(project_and_group)
+        job_token_scope = project.job_token_scope.get()
+
+        self._process_limit_access_to_this_project_setting(
+            job_token_config, job_token_scope
+        )
+
+        job_token_scope.refresh()
+
+        allowlist_config = job_token_config.get("allowlist", {})
+        verbose(f"configuration allowlist: {allowlist_config}")
+
+        if allowlist_config:
+            verbose("Processing Job Token allowlist")
+
+            self._process_groups(job_token_scope, allowlist_config.get("groups", []))
+            job_token_scope.refresh()
+            self._process_projects(
+                project, job_token_scope, allowlist_config.get("projects", [])
+            )
+        else:
+            warning(
+                "Removing existing groups and projects from Allowlist, as none set in configuration"
+            )
+
+            self._remove_groups_from_allowlist(
+                job_token_scope, job_token_scope.groups_allowlist.list(), []
+            )
+            job_token_scope.refresh()
+            self._remove_projects_from_allowlist(
+                project, job_token_scope, job_token_scope.allowlist.list(), []
+            )
+
+        job_token_scope.refresh()
+
+    @staticmethod
+    def _process_limit_access_to_this_project_setting(
+        configuration: dict, job_token_scope: ProjectJobTokenScope
+    ):
+        limit_access_to_this_project: bool = configuration.get(
+            "limit_access_to_this_project", True
+        )
+        verbose(f"limit_access_to_this_project: {limit_access_to_this_project}")
+        job_token_scope.enabled = limit_access_to_this_project
+        job_token_scope.save()
+
+    def _process_projects(
+        self,
+        project: Project,
+        job_token_scope: ProjectJobTokenScope,
+        projects_allowlist: List,
+    ):
+        if not projects_allowlist:
+            warning(
+                "Process will remove existing projects from allowlist, as none set in configuration"
+            )
+
+        existing_allowlist = job_token_scope.allowlist.list()
+
+        target_project_ids = self._get_target_project_ids_from_config(
+            projects_allowlist
+        )
+
+        project_ids_to_add = self._get_ids_to_add_to_allowlist(
+            existing_allowlist, target_project_ids
+        )
+        for project_id in project_ids_to_add:
+            if project_id != project.id:
+                job_token_scope.allowlist.create({"target_project_id": project_id})
+                job_token_scope.save()
+
+        self._remove_projects_from_allowlist(
+            project, job_token_scope, existing_allowlist, target_project_ids
+        )
```

Comment:
Is this being done twice? Once here and then again on line 47?

Reply:
line 47 catches if there is no "allowlist" specific in the configuration for example if a user went from
```
        projects_and_groups:
          proj:
            job_token_scope:
              limit_access_to_this_project: true
              allowlist:
                groups:
                  - group-a
```
to
```
        projects_and_groups:
          proj:
            job_token_scope:
              limit_access_to_this_project: true
```
 We would want to remove group-a from the allow list, without having to go off and get the existing allowlist and compare to the one defined in config (empty)